src/Autofac/Core/Lifetime/LifetimeScope.cs

```diff
@@ -48,7 +48,7 @@ public class LifetimeScope : Disposable, ISharingLifetimeScope, IServiceProvider
         /// Protects shared instances from concurrent access. Other members and the base class are threadsafe.
         /// </summary>
         private readonly object _synchRoot = new object();
-        private readonly ConcurrentDictionary<Guid, object> _sharedInstances = new ConcurrentDictionary<Guid, object>();
+        private readonly ConcurrentDictionary<(Guid, Guid?), object> _sharedInstances = new ConcurrentDictionary<(Guid, Guid?), object>();
```

Comment:
Made the latter value nullable, since defaulting to Guid.Empty had a negative impact to a TON of unit tests, triggering a condition that shouldn't have been triggered around self-registration.

Reply:
I'm wondering, if the additional 300ns perf cost is down to the HashCode.Combine and additional equivalence checks in the dictionary, does it make sense to have two ConcurrentDictionary instances? One with a single-guid key, one with a double? Use the single if the qualifying ID is null? That might remove the additional cost for any non-decorator paths.

Reply:
I thought the same thing, mostly in the instantiation of the struct plus the more complicated hashcode generation, since I *think* Dictionary<> uses the hashcode for equivalence on keys. The split sounds like a good path for performance, but does introduce a bit of extra maintenance complexity that could grow until the point that the pipeline stuff comes to fruition. Performance wins out, in this scenario, though!